libminifi/test/TestBase.cpp

```diff
@@ -62,78 +63,51 @@ TestPlan::~TestPlan() {
   for (auto& processor : configured_processors_) {
     processor->setScheduledState(core::ScheduledState::STOPPED);
   }
+  for (auto& connection : relationships_) {
+    // This is a patch solving circular references between processors and connections
+    connection->setSource(nullptr);
+    connection->setDestination(nullptr);
+  }
   controller_services_provider_->clearControllerServices();
 }
 
 std::shared_ptr<core::Processor> TestPlan::addProcessor(const std::shared_ptr<core::Processor> &processor, const std::string &name, const std::initializer_list<core::Relationship>& relationships,
-                                                        bool linkToPrevious) {
+    bool linkToPrevious) {
   if (finalized) {
     return nullptr;
   }
   std::lock_guard<std::recursive_mutex> guard(mutex);
-
   utils::Identifier uuid = utils::IdGenerator::getIdGenerator()->generate();
-
   processor->setStreamFactory(stream_factory);
   // initialize the processor
   processor->initialize();
   processor->setFlowIdentifier(flow_version_->getFlowIdentifier());
-
   processor_mapping_[processor->getUUID()] = processor;
-
   if (!linkToPrevious) {
     termination_ = *(relationships.begin());
   } else {
     std::shared_ptr<core::Processor> last = processor_queue_.back();
-
     if (last == nullptr) {
       last = processor;
       termination_ = *(relationships.begin());
     }
-
-    std::stringstream connection_name;
-    connection_name << last->getUUIDStr() << "-to-" << processor->getUUIDStr();
-    logger_->log_info("Creating %s connection for proc %d", connection_name.str(), processor_queue_.size() + 1);
-    std::shared_ptr<minifi::Connection> connection = std::make_shared<minifi::Connection>(flow_repo_, content_repo_, connection_name.str());
-
     for (const auto& relationship : relationships) {
-      connection->addRelationship(relationship);
-    }
-
-    // link the connections so that we can test results at the end for this
-    connection->setSource(last);
-    connection->setDestination(processor);
-
-    connection->setSourceUUID(last->getUUID());
-    connection->setDestinationUUID(processor->getUUID());
-    last->addConnection(connection);
-    if (last != processor) {
-      processor->addConnection(connection);
+      addConnection(last, relationship, processor);
```

Comment:
We used to have a single Connection with multiple Relationships, now we'll have a separate Connection for each Relationship.  What is the reason for this change?

Reply:
I am not sure I understand the question, but It is a prerequisite if you want to connect up multiple relationships separately.

Reply:
The behavior has changed between the old code and the new code.  Are you saying the old code was broken, and this fixes the problem?  If so, what was the problem?

Reply:
I think this is the same behavior, just extracted to a separate function.

Reply:
There is a difference between 1 connection for 3 relationships and 3 connections for 3 relationships. The latter may or may not be fine, so it's better not to change the behavior without a good reason.

Reply:
Why do you expect behavioural differences between the two? If you want to have more ellaborate test-setups, you need to have the flexibility of separately linking up relationships on processors.

From a more pragmatical standpoint, this change is also tested through all the unit tests ran.

Reply:
This change was reverted in https://github.com/apache/nifi-minifi-cpp/pull/940/commits/33e84f4f31b803a0cf0f212287c701d922295059.